bundle/permissions/workspace_root.go

```diff
@@ -52,16 +55,58 @@ func giveAccessForWorkspaceRoot(ctx context.Context, b *bundle.Bundle) error {
 	}
 
 	w := b.WorkspaceClient().Workspace
-	obj, err := w.GetStatusByPath(ctx, b.Config.Workspace.RootPath)
+	rootPath := b.Config.Workspace.RootPath
+	paths := []string{}
+	if !libraries.IsVolumesPath(rootPath) {
+		paths = append(paths, rootPath)
+	}
+
+	if !strings.HasSuffix(rootPath, "/") {
+		rootPath += "/"
+	}
+
+	if !strings.HasPrefix(b.Config.Workspace.ArtifactPath, rootPath) &&
+		!libraries.IsVolumesPath(b.Config.Workspace.ArtifactPath) {
+		paths = append(paths, b.Config.Workspace.ArtifactPath)
+	}
+
+	if !strings.HasPrefix(b.Config.Workspace.FilePath, rootPath) &&
+		!libraries.IsVolumesPath(b.Config.Workspace.FilePath) {
+		paths = append(paths, b.Config.Workspace.FilePath)
+	}
+
+	if !strings.HasPrefix(b.Config.Workspace.StatePath, rootPath) &&
+		!libraries.IsVolumesPath(b.Config.Workspace.StatePath) {
+		paths = append(paths, b.Config.Workspace.StatePath)
+	}
+
+	if !strings.HasPrefix(b.Config.Workspace.ResourcePath, rootPath) &&
+		!libraries.IsVolumesPath(b.Config.Workspace.ResourcePath) {
+		paths = append(paths, b.Config.Workspace.ResourcePath)
+	}
+
+	g, ctx := errgroup.WithContext(ctx)
+	for _, p := range paths {
+		g.Go(func() error {
+			return setPermissions(ctx, w, p, permissions)
+		})
+	}
+
+	return g.Wait()
+}
+
+func setPermissions(ctx context.Context, w workspace.WorkspaceInterface, path string, permissions []workspace.WorkspaceObjectAccessControlRequest) error {
+	obj, err := w.GetStatusByPath(ctx, path)
```

Comment:
Is it guaranteed that all paths exist at this point?

If not, then this function should do a create-if-not-exists.

Reply:
It's guaranteed because it's done after files.Upload

Reply:
@andrewnester Can you clarify this one?

E.g. does the state path or the resources path exist after `files.Upload`?

Reply:
Yes, all of them will be created prior, the current mutator order is

```
  libraries.Upload(), // creates artifacts path
  ...
  files.Upload(outputHandler), // creates files and resources path
  ...
  deploy.StatePush(), // creates state path
  permissions.ApplyWorkspaceRootPermissions(),
```